django/db/models/sql/compiler.py

```diff
@@ -1768,12 +1785,23 @@ def as_sql(self):
 
         placeholder_rows, param_rows = self.assemble_as_sql(fields, value_rows)
 
-        on_conflict_suffix_sql = self.connection.ops.on_conflict_suffix_sql(
+        (
+            on_conflict_suffix_sql,
+            where_clause,
+        ) = self.connection.ops.on_conflict_suffix_sql(
             fields,
             self.query.on_conflict,
             (f.column for f in self.query.update_fields),
             (f.column for f in self.query.unique_fields),
+            where_clause=self.query.where_clause,
```

Comment:
I've realised that we're passing this around but not actually doing anything with it in the backend-specific operations. I had envisaged that we'd actually add the `WHERE ... ` part to the SQL inside `on_conflict_suffix_sql()` as it may be that the different backends required different placement... 🤔 It'd be good if you could comment with the details of what the different backends support here so that we can see, but it'd still be good to push down the `compile_where_clause_to_sql()` thing if we can or it makes sense.

I probably need to give this some more thought which I won't get around to for a while, so feel free to have a play around.

Reply:
i've added the WHERE ... constructor inside on_conflict_suffix_sql to move the compile_where_clause_to_sql function into the base feature. it's now further down.